### DISCUSSION TITLE
2024-10-04 - GitHub Copilot Chat in GitHub.com - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by bradystroud** [SSWConsulting/SSW.Rules.Content#9339] [Chewing the Fat]
Hey SSWers,

This week we are looking at GitHub Copilot Chat in GitHub.com

https://www.ssw.com.au/rules/ai-pair-programming/

What is GitHub Copilot's strong suit? 💪
What is GitHub Copilot's weak spot? 😔

**Comment by JeanThirion** [comment]
💪The abilty to answer contextual question on the repo + wiki and give precise reference is really good. We've all been looking at code forever trying to work it out, and wishing we had an expert available to guide us.

😔 Automated comments/PR descriptions is to me quite bad right now... It's very verbose and if it's just to describe what's going on without any smarts, it's just an empty shell. I wouldn't want devs to fill a PR description just for the sake of it, it has to have meaning and reasoning behind it.

**Comment by brookjeynes-ssw** [comment]
I haven't used GitHub Copilot recently. The last time I used it however, its strong suit was writing repetitive code while its weak spot is that it can often get in the way of what I actually want to do.

**Comment by drwharris** [comment]
💪I am not really sure as I dont use it. From the video, it appears to be good at pointing you to where certain behaviours occur in a code base
😔From my minimal use of Copilot, it is overly verbose, I always have to clean up anything I do use and it tended (as most AI seems to) miss key chunks of code areas to test.

**Comment by jakebayliss** [comment]
💪 Awesome that we can ask questions about our code! Generating very basic methods/classes, refactoring code. 
😔  I have found trying to generate tests hasn't been great. Always have to fix things. But it is getting a lot better

**Comment by TobyChurches** [comment]
> What is GitHub Copilot's strong suit? 💪

Great for boilerplate code generation + suggestions that make sense in the context. I sometimes just write the comment first and Copilot will do the rest

> What is GitHub Copilot's weak spot? 😔

Can sometimes be a bit slow to generate the suggestions + the comments can generally be a bit verbose

**Comment by ncn-ssw** [comment]
> What is GitHub Copilot's strong suit? 💪

💪 Boilerplate - quickly completes repetitive code
💪 Patterns - if a pattern is common enough, it can generate a whole function
💪 Guidance - the explanation generation can be useful when learning a new stack, framework or language

> What is GitHub Copilot's weak spot? 😔

😔 Normally wrong - unfortunately, it more often generates code to do stuff I don't want it to do. I could function without it, but I want to keep an eye on how the technology develops.

**Comment by duanxinhuan** [comment]
ChatGPT works fine for me on most of the cases, that's why I do not use copilot.

**Comment by kikibianc** [comment]
As a SysAdmin, very little use of Copilot - ChatGPT is generally the way to go, as I am not in GitHub the whole day.
If it was free and just there, then I would definitely use more.

**Comment by joshbermanssw** [comment]
Have recently moved away from CoPilot but 
💪 Repetitive code 
💪 General Explanation of code 
😔 Might take one or two extra prompts to get it right

**Comment by SamBWagner** [comment]
> What is GitHub Copilot's strong suit? 💪

It's pretty quick and easy to use, especially if you're using VSCode. Can make boilerplate less annoying to write.

> What is GitHub Copilot's weak spot? 😔

I tend to find that it hallucinates a lot and I spend a large portion of my time just fixing it's mistakes.

**Comment by AntPolkanov** [comment]
> What is GitHub Copilot's strong suit? 💪

- Saves time when code typing
- I like the idea that it can analyse a repository and help you understand it

> What is GitHub Copilot's weak spot? 😔

- Your brain might not keep up with what Copilot is generating
- Developers may become lazy

Personally, I prefer to use ChatGPT together with the [Full Line Code Completion](https://plugins.jetbrains.com/plugin/14823-full-line-code-completion) plugin in Rider. The plugin generates code line-by-line, so you still have that feeling that you are in the driver seat. I really enjoy working with it.

**Comment by leahy268** [comment]
>What is GitHub Copilot's strong suit? 💪
>What is GitHub Copilot's weak spot? 😔
I don't think I've really used it enough to understand this

**Comment by ricksu978** [comment]
> What is GitHub Copilot's strong suit? 💪

It's quite good at picking up the design pattern I have used in the project.

> What is GitHub Copilot's weak spot? 😔

It's not good at making suggestions for less popular language, e.g. bicep

**Comment by bettybondoc** [comment]
N/A - I'm not a developer

**Comment by RobJT-wq** [comment]
As a Sysadmin, I don't live in GitHub but I can see how it would be useful to assist with writing code.

**Comment by calumjs** [comment]
💪Being integrated in GitHub is cool, eventually you will be able to create an issue and it will just suggest a solution for you
😔I can't use Claude with GitHub Copilot - I would rather use Cursor, or just Claude directly

**Reply by calumjs** [reply]
Check out Cursor if you are interested - it has a free option https://www.cursor.com/

**Comment by steven0x51** [comment]
💪 understand the context of the code so able to provide a better solution than other AI, And it can be embedded into an IDE

😔 no free tier

**Comment by jeoffreyfischer** [comment]
**strong suit** - accessibility: it can easily be used straight in the IDE
**weak spot** - not always accurate: I've had suggestions that were wrong

**Comment by Calinator444** [comment]
> What is GitHub Copilot's strong suit? 💪

- Great for drafting up proof of concepts when defining new functions
- Much quicker than using stack overflow
- Surprisingly accurate

> What is GitHub Copilot's weak spot? 😔

- It's certainly not a replacement for a senior developer (maybe someday it will be but it's not there yet)
- It's important that you understand the code that copilot wrote, particularly if you don't have unit tests
- Take any domain specific knowledge it provides with a grain of salt (i.e. whether or not you should include typescript in your project)
- CAREFULLY TEST THE CODE IT PRODUCES. I've had occasions where copilot has referred to outdated documentation (i.e. only the documentation available when the LLM was trained) or even make up methods and attributes for libraries. The user documentation should be trusted over something that copilot said.

**Comment by chrisschultzssw** [comment]
N/A - not a dev :).

😔 [AI is helping write code for hackers](https://www.bleepingcomputer.com/news/security/hackers-deploy-ai-written-malware-in-targeted-attacks/) 🙃

**Comment by tkapa** [comment]
I've had Copilot for about a year, and unfortunately I can probably count on one hand the number of times it's been helpful.

Those "Thanks Copilot" moments are few and far between, but maybe I just don't know how to use it.

**Reply by Calinator444** [reply]
Weird flex but ok

**Comment by awaltos** [comment]
💪️ Helps to understand code using simple language
😔 Still not always accurate. Some responses are just flat out wrong and completely made up

**Comment by edkranz** [comment]
💪 Writing any code which follows common patterns.

😔 It doesn't actually know how to think, so if you are trying to write something that hasn't been written before, it will struggle (and likely fail).
Although the newer slow AI models (e.g. o1), might be able to 'think' about novel problems and fix this problem. We'll see.

**Reply by Calinator444** [reply]
Something I'd be really interested to see is the option to switch between language models in Copilot. From what I've heard Claude is much better for code generation than ChatGPT 4o.

**Comment by bahjatalaadelssw** [comment]
I don't if it's better than ChatGPT or if it follows clean architecture principles.

**Comment by stephanfakossw** [comment]
N/A - I'm not a developer 
![ErrorGIF](https://github.com/user-attachments/assets/3f7453a5-9c3a-49aa-9252-920b1b2870e8)

**Comment by sam-s-maher** [comment]
I don't use it yet, but I plan to!

Seems like the consensus is clear though, it's an efficient code generator, not necessarily and effective one.

**Comment by MattParkerDev** [comment]
Strong: helping you with syntax you have forgotten, and giving you a good starting point on a task you may not have any idea how to do.

Weak: You need to be experienced enough to understand the code that it is writing, so you can know when what it just wrote is rubbish, or if it needs tweaking.

**Comment by wicksipedia** [comment]
tl;dr - it's a tool and you can footgun with it

**Strong suit:** `/explain` can break down complex code or piped shell commands

**Weak spot:** since it uses the existing codebase as context, it is more likely to generate a bad response by not suggesting something out of the box. For example, on HW, I needed to calculate whether a polyline was intersecting with polygons on GoogleMaps; the GoogleMaps SDK has some methods to let you do this, and Copilot was suggesting to use them but after reading the docs - they wouldn't work in the same way that I needed them to. There was an alternative library that could do the calcs as required; which ChatGPT suggested since it didn't have the same context

**Comment by amankumarrr** [comment]
💪 Helps to understand complex code
💪 Helps to write better code with suggestions

😔 Requires some back-and-forth to get the desired result
😔 Lacks the human touch that ChatGPT provides

**Comment by Ben0189** [comment]
>What is GitHub Copilot's strong suit? 💪

I like how you can use copilot to ask question on the repo 

>What is GitHub Copilot's weak spot? 😔

I have use it a couple of time and it gives bad response🥲

What about Claude 3.5 sonnet https://www.anthropic.com/news/claude-3-5-sonnet

**Comment by StefStar** [comment]
N/A Not a developer :((

**Comment by jimmy1992abc** [comment]
I am not a developer, but I have had many chances to interact with Copilot (from the other MS services, e.g., Power Apps).
> What is GitHub Copilot's strong suit?
Good integration within the software or environment I am using.

> What is GitHub Copilot's weak spot?
It doesn't fix my problem. 
It feels not smart enough and cannot understand what I want to achieve. The code it changed for me is not working.

**Comment by jernejk** [comment]
💪 Is in the background helping me when I needed
😔 Sometimes it stops working (usually need to re-auth or restart VS) and I don't notice for days until I need a specific solution. To be fair, I use ChatGPT for tough problems.

PS: For anyone who tried this 1-2 months ago, now it actually works. Before, it couldn't answer a single question other than what was in the description of repo.

**Comment by AttackOnMorty** [comment]
> What is GitHub Copilot's strong suit? 💪

It's my silent driver in pair programming. 
- Wirte unit tests
- Explain code

> What is GitHub Copilot's weak spot? 😔

Generated code is sometimes unreliable, but you can write tests to verify that.

**Comment by JackDevAU** [comment]
💪 Writing simple code e.g. json objects and simple functions. 
😞 Writing anything complex or something it doesn't have context/isn't trained on

**Comment by Marxoz** [comment]
> What is GitHub Copilot's strong suit? 💪

Being able to build its own context about the solution by simply reading the code

   > What is GitHub Copilot's weak spot? 😔

No context might yield the wrong results. Maybe adding sufficient information to the ReadMe might mitigate this issue.

Does this mean that code comments are back???

**Comment by Tanya-Leahy** [comment]
N/A - I'm not a developer

**Comment by GordonBeeming** [comment]
> What is GitHub Copilot's strong suit? 💪

Gives you ideas for code/text that you might not have even been thinking about

> What is GitHub Copilot's weak spot? 😔

Sometimes the suggestions are a distraction (when occasionally they aren't too helpful)

**Comment by bradystroud** [comment]
💪 Explaining logic
💪 Rubber ducking 🦆 
💪 Alternate perspective
💪 Boilerplate code
💪 Brainstorming
💪 Remembering syntax things (I know what I want, but not the exact syntax to do it)

😔 Zooming out e.g. Is this the right place for this code?
😔 Up to date knowledge - it still doesn't like using file-scoped namespaces and other newer c# features

**Comment by matt-goldman-ssw** [comment]
Strong suit - it's excellent at inferring context from your open files
Weak spot - chat is not very good, but looks like it's getting better based on that video

**Comment by babakamyljanovssw** [comment]
I usually use Copilot when I need to understand or break down a piece of code that I'm not familiar with or generating boilerplate.

    > 1. What is GitHub Copilot's strong suit? 💪
Explains piece of code quickly
Generating boilerplate code
Generating commit messages and pull requests summaries

    > 2. What is GitHub Copilot's weak spot? 😔
Can generate solution/code with the bug or not the thing I want, so it takes more tries to get what I want

**Comment by vladislav-kir** [comment]
> What is GitHub Copilot's strong suit? 💪

Explaining code & Generating boilerplate code

> What is GitHub Copilot's weak spot? 😔

Generating code that is more complex.

**Comment by tino-liu** [comment]
> What is GitHub Copilot's strong suit? 💪

Generating boilerplate and explaining code. Also some widely used algos 😜

> What is GitHub Copilot's weak spot? 😔

Context-based code

**Comment by PennyWalker** [comment]
Can't really comment on this one because I'm not a dev, but I can see how useful this can be for the more simple stuff.

I love getting my HTML or Markdown from ChatGPT :D

**Comment by ozair-ashfaque** [comment]
> What is GitHub Copilot's strong suit? 💪

It has the context of the codebase.

> What is GitHub Copilot's weak spot? 😔

Sometimes it gives wrong answers like ChatGPT.

**Comment by danielmackay** [comment]
> What is GitHub Copilot's strong suit? 💪

Still seems to be one of the top AI programmers (Codeium is not far behind though).  Deep integration with GitHub

> What is GitHub Copilot's weak spot? 😔

Integration with Rider is not as tight as VS and VS Code (although that's probably on purpose).  For a personal licence the cost is still high for what you get.

**Comment by jerryluossw** [comment]
> What is GitHub Copilot's strong suit? 💪

- Generated code is very practical if the description is accurate.
- explaining code

 > What is GitHub Copilot's weak spot? 😔

- Sometimes it's lag when using it in Visual Strudio.

**Comment by Aibono1225** [comment]
What is GitHub Copilot's strong suit? 💪
- Explaning code, this is so cool
- Auto commit message
- Write repetitive code

What is GitHub Copilot's weak spot? 😔
- Code generated can sometimes (or mostly) be problematic, needing further improvement
